[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90801

Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

jcaip committed:
[pruning][core][feature] LSTM Structured Pruning

Summary:

TODO: Fill this in

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning pattern + prune function"


Summary:

TODO: Fill this in

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on test/ao/sparsity/test_structured_sparsifier.py:
```diff
@@ -639,3 +663,92 @@ def test_complex_conv2d(self):
+        _, _ = pruned_model(rnn_input)
+
+        expected_params = dict(model.named_parameters())
+        for name, param in model.named_parameters():
```
If anyone has a better idea about how to test this, please let me know.

WongJohnson replied:
It appears you are testing the parameters against themselves there? Do you mean to iterate over the pruned_model?

WongJohnson replied:
If so, I believe this is equivalent to your test:
```
pruned_params = dict(pruned_model.named_parameters())
for name, param in model.named_parameters():
    assert name in pruned_params, f"pruned param {name} is missing"
    assert elements_are_subset(pruned_params[name], param), f"pruned param {name} ({pruned_params[name].shape}) is not a subset of model ({param.shape})"
```

jcaip replied:
thanks, good catch!

WongJohnson commented on torch/testing/_internal/common_pruning.py:
```diff
@@ -7,6 +7,10 @@
 from torch import nn
 
 
+def elements_are_subset(subset_tensor, superset_tensor):
```
I didn't follow the rest of your code, but I assume pruning here means that certain "rows" are simply removed from the superset_tensor. If that's the case, you can search for the `subset_tensor` as you sequentially sweep through `superset_tensor` to avoid a lot of duplicate comparisons (and also avoid some edge case false True returns, i.e. matching the same row multiple times):

```
def elements_are_subset_ordered(subset_tensor, superset_tensor):
    i = 0
    for row in subset_tensor:  # or iterate over whichever dimension is appropriate
        while i < len(superset_tensor):
            if not torch.equal(row, superset_tensor[i]):
                i += 1  # keep searching
            else:
                break  # check next row (continue for loop)
        else:
            return False  # searched beyond end of superset_tensor

    return True  # matched all in subset_tensor
```

I'm happy to discuss if I misunderstood the assumptions here.

Commit:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot rebase fbcode/warm

jcaip commented:
@pytorchbot --help

jcaip commented:
@pytorchbot --help

jcaip commented:
@pytorchbot rebase -b fbcode/warm

jcaip commented:
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

jcaip commented:
@pytorchbot rebase -b fbcode/warm

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip commented:
@pytorchbot rebase

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning.

It works by splitting the packed weights, pruning each piece individually, and then stichin back together the results.
Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning

Adds in test to check that LSTM pruning works

Added in support for multi-layer LSTMs

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip commented:
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

jcaip commented on test/ao/sparsity/test_structured_sparsifier.py:
```diff
@@ -75,6 +95,72 @@ def test_update_mask(self):
         assert expected.shape == pruned.shape
         assert torch.isclose(expected, pruned, rtol=1e-05, atol=1e-07).all()
 
+    def test_lstm_saliency_pruner_update_mask(self):
```
@WongJohnson I added this test for the SaliencyPruner and  two additional tests for LSTM -> layernorm -> Linear below.

WongJohnson replied:
Great, thanks for adding this. I'd like to see a bit more complex tests (perhaps randomly-drawn tensors) but I'll leave that up to your team's judgement. I would at least ensure that you test properly handling negative weights and also 1d tensors (i.e. bias tensors for which I had to add the dims() check - I believe `SaliencyPruner` still has this issue).

jcaip replied:
I think actually pruning 1d tensors is something that should not work - since we do rowise pruning, it doesn't really make sense to prune a 1d tensor. When we prune the bias, we're actually just taking the mask from the weight and applying it to the bias tensor, not computing a mask for the bias itself. This should probably throw a clearer error message earlier though.

WongJohnson commented on torch/testing/_internal/common_pruning.py:
```diff
@@ -8,7 +8,16 @@
+                break
+        else:
+            return False
+    return True
```
Some comments regarding assumptions (i.e. subset and order preserved) would help future use of this function, since it no longer matches the function name.

WongJohnson reviewed (commented):
Thanks for incorporating some of my suggestions.

WongJohnson commented on torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:
```diff
@@ -0,0 +1,36 @@
+                mask = cast(torch.Tensor, p.mask)
+
+                # select weights based on magnitude
+                if weights.dim() > 1:
+                    # take norm over all but first dim
+                    dims = tuple(range(1, weights.dim()))
+                    saliency = weights.abs().norm(dim=dims, p=1)
+                else:
+                    # 1d param: use weights directly
+                    saliency = weights.abs()
+
+                # handle weights in 4 groups
+                split_size = len(mask) // 4
+                masks = torch.split(mask, split_size)
+                saliencies = torch.split(saliency, split_size)
+
+                for keep_mask, sal in zip(masks, saliencies):
+                    # mask smallest k values to be removed
+                    k = int(len(keep_mask) * kwargs["sparsity_level"])
+                    prune = sal.topk(k, largest=False, sorted=False).indices
```
Perhaps pull this logic out (minus the 4-split) and share between `SaliencyPruner` and `LSTMSaliencyPruner`

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip commented:
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

jerryzh168 commented on torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:
```diff
@@ -99,6 +103,10 @@ def _get_default_structured_pruning_patterns() -> Dict[
         (nn.Conv2d, "output"): prune_conv2d,
         (nn.Conv2d, nn.Conv2d): prune_conv2d_conv2d,
+        # lstm -> getitem -> linear
+        (nn.LSTM, getitem, nn.Linear): prune_lstm_linear,
```
I remember the output of LSTM is `output, (h_n, c_n)`, so there will be multiple getitem node after the call to forward, it won't affect the current use case much probably since getitem is followed by some other ops, but this is not very robust I think

jcaip replied:
Yeah, this is very brittle - I tried to match `getitem(0)` instead of `getitem` only but couldn't figure out how to do that, if you have know how?

I can rename this to `prune_lstm_output_linear` I think to make it a bit more clear it's only for lstm -> output -> linear and not lstm -> h_n / c_n -> linear. The pruning function implementation is only correct lstm-> output -> linear anyways, it applies the wrong mask for h_n / c_n.

jerryzh168 commented on torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:
```diff
@@ -0,0 +1,34 @@
+                saliency = weights.norm(dim=dims, p=1)
+
+                # handle weights in 4 groups
+                split_size = len(mask) // 4
```
why do we handle weights in 4 groups?

WongJohnson replied:
There are four sets of weights in each tensor. It may be worth verifying whether they should be pruned completely independently.
https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#:~:text=of%20shape%20(-,4*hidden_size,-%2C%20input_size)%20for
(search for "4*hidden_size" if the link doesn't jump for you)

jerryzh168 replied:
I see, thanks, would be helpful to add a comment I think

WongJohnson commented on torch/testing/_internal/common_pruning.py:
```diff
@@ -7,6 +7,22 @@
 
+def rows_are_subset(subset_tensor, superset_tensor) -> bool:
+    """
+    Checks to see if all rows in subset tensor are present in the superset tensor
```
Preserved ordering of the subset tensor is the second assumption I was referring to - that is the basis of the speed-up compared to the original implementation.

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip commented:
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

jerryzh168 commented on torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:
```diff
@@ -83,14 +87,14 @@ def _get_supported_activation_modules():
     """
     patterns: Dict[
-        Tuple[Union[Type[nn.Module], Callable[[torch.Tensor], torch.Tensor], str], ...],
+        Tuple[Union[Type[nn.Module], Callable, MatchAllNode, str], ...],
```
is this change needed? didn't see MatchAllNode being used

jerryzh168 reviewed (approved):
LGTM, but maybe @WongJohnson should take a final look since I'm not very familiar with  details

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot rebase

jcaip committed:
Update on "[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern"


Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge
